### PR TITLE
use more elements per thread for narrow dtypes

### PR DESCRIPTION
**Pull request by ngimel**
Fix perf issue for narrow type by accessing more elements per thread

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by Chillee — APPROVED** [review]
How do the benchmarks look?

**Comment by ngimel** [comment]
It's on par with wider types now

**Comment by eqy** [comment]
Somehow this broke bazel? Maybe https://github.com/pytorch/pytorch/blob/6ef6b3f586c6f7b06c6b3b1bf7c264cc7f044687/tools/rules_cc/cuda_support.patch#L18 needs to be updated?

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 1]
```diff
@@ -52,13 +52,49 @@
...
+
+template <int io_sizes>
+constexpr auto elems_per_thread(){
+  if constexpr (io_sizes == 1) {
+    return 16;
+  } else if constexpr (io_sizes < 4) {
+    return 8;
+  } else {
+    return 4;
+  }
```
Comment (eqy) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:73]: nit: can this be expressed as something like `bytes_per_thread//size`?
Just thinking about this potentially being potentially hw-dependent where we might want different bytes per thread

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/CUDALoops.cuh (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:73]: For ease of reasoning it's better if it's power of 2, and `bytest_per_thread//size` won't have this property. Besides, for large sizes we still probably want to keep 4, it served us well all these years.

**Comment by ngimel** [comment]
@huydhn it was a legit test failure, do you know why the test didn't run on the PR?

**Pull request closed without merging**